gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/metrics/kafka/KafkaAvroEventKeyValueReporter.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.metrics.kafka;
+
+import avro.shaded.com.google.common.base.Optional;
+import avro.shaded.com.google.common.collect.Lists;
+import com.google.common.base.Splitter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.gobblin.metrics.GobblinTrackingEvent;
+import org.apache.gobblin.metrics.MetricContext;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+import java.util.Queue;
+
+/**
+ * {@link org.apache.gobblin.metrics.reporter.EventReporter} that emits events to Kafka as serialized Avro records with a key.
+ * Key for these kafka messages is obtained from values of properties provided via {@link #keyPropertyName}.
+ * If the GobblinTrackingEvent does not contain any of the required property, key is set to null. In that case, this reporter
+ * will act like a {@link org.apache.gobblin.metrics.kafka.KafkaAvroEventReporter}
+ */
+@Slf4j
+public class KafkaAvroEventKeyValueReporter extends KafkaAvroEventReporter {
```

Comment:
If you want to reuse this logic for other reporter types like KafkaJsonEventKeyValueReporter then this should extend the KafkaEventReporter.